### DISCUSSION TITLE
Will it be difficult to write a parser for Robot Framework?

### DISCUSSION THREAD

**Discussion by Hubro** [tree-sitter/tree-sitter#1724] [Q&A] [answered]
I work a lot with Robot files these days, and I'm really annoyed by the terrible syntax highlighting I get in Neovim. I want to write a tree-sitter parser for it.

However, a while back I wrote a parser for YANG, and while researching how to do that, I got the sense that writing parsers for languages where white space indentation is significant (such as Python) is really hard.

In Robot files, white space is even more significant than it is in Python. Not only is indentation significant, but the white space between words are as well. A single space is considered part of a keyword. For example, `Open A New Tab` is a single keyword, whereas `Open Tab    address=foo` (three spaces before `address`, but GitHub strips them) is a keyword with an argument. In other words, tokens are separated by *two or more* spaces or one or more tab stops.

Does this kind of language necessitate using C to parse the white-space, like Python does? Has tree-sitter gained better support for this since last summer?

Here are a couple of example Robot files, for reference:

- https://github.com/robotframework/RobotDemo/blob/master/keyword_driven.robot
- https://github.com/robotframework/RobotDemo/blob/master/data_driven.robot

**Comment by ahelwer** [comment, marked as answer]
You don't necessarily need to use an external scanner to parse whitespace. Python needs to do that because it has to keep track of the column position for indented code blocks across multiple lines. As you've described it it's conceivable you could parse this without writing an external scanner. It might get a bit tricky with the extras field and discerning which whitespace is and is not syntactically relevant, but hard to say until you try.

**Reply by Hubro** [reply]
Alright, that's encouraging. I think I'll give it a try as soon as I can find some free time.
